### DISCUSSION TITLE
GSoC 2024: List of Ideas

### DISCUSSION THREAD

**Discussion by ashitaprasad** [foss42/apidash#112] [General]
GSoC 2024 [ **List of Ideas** | [Application Guide](https://github.com/foss42/apidash/discussions/111) ]

🚨 We are participating in GSoC 2024 🎉 

<img src="https://github.com/foss42/apidash/assets/615622/493ce57f-06c3-4789-b7ae-9fa63bca8183" alt="GSoC" width="500">

| | Link |
|--|--|
| Learn about GSoC | [Link](https://summerofcode.withgoogle.com) |
| Organization page on GSoC | [Link](https://summerofcode.withgoogle.com/programs/2024/organizations/api-dash) |
| Project Ideas List | [Link](https://github.com/foss42/apidash/discussions/112) |
| Application Guide | [Link](https://github.com/foss42/apidash/discussions/111) |
| Discord Channel | [Link](https://discord.com/invite/2s49SCNfyJ) |

The objective of API Dash project is to ease the exploration, testing & integration of APIs. These ideas have been discussed with all the mentors in our team:
- [Ashita Prasad](https://github.com/ashitaprasad)
- [Ankit Mahato](https://github.com/animator)

The list of ideas provided below are not restrictive. In case you have any other innovative idea in mind which can help fellow developers in tackling the various problems of the API Lifecycle, please feel free to comment below and [open a new issue for discussion](https://github.com/foss42/apidash/issues/new/choose). 

If a project is successfully selected you will be allocated a primary mentor and supported by the entire team. If you are interested in learning more about a particular idea please communicate in the corresponding "related issue" thread or feel free to drop by our [Discord server](https://bit.ly/heyfoss) and we can have a chat.

### 1. Android/iOS support 
Related Issue - https://github.com/foss42/apidash/issues/33 

API Dash is currently available across all three desktop platforms - Windows, macOS, Linux. One of the reasons we were able to target these platforms was due to the choice of our cross-platform development framework - Flutter. We are looking forward to adding support for mobile platforms (Android, iOS, iPad OS). It would provide agility and enable everyone to test APIs on-the-go, having the same familiar UX and set of features providing continuity. And everything will be powered by a single codebase. The task will also involve writing integration tests to test API Dash workflows on mobile.

**Skills**: UI/UX Design, Dart, Flutter, Testing
**Difficulty**: Medium-High
**Length**: 175 hours

### 2. Realtime API Testing Support - WebSocket, SSE, MQTT 
Related Issue - https://github.com/foss42/apidash/issues/15 https://github.com/foss42/apidash/issues/115 https://github.com/foss42/apidash/issues/116

Testing WebSocket, MQTT (Message Queuing Telemetry Transport), and SSE (Server-Sent Events) protocols is crucial for ensuring the reliability, scalability, and security of real-time communication systems. Each of these protocols serves different purposes and is utilized in various applications ranging from finance to web applications to IoT (Internet of Things) devices. The objective of this project is to design the architecture of the core library, understand the specs & implement the support for testing, visualization & integration code generation of Realtime APIs in API Dash.

**Skills**: Understanding Specs, Dart, Flutter
**Difficulty**: Medium-High
**Length**: 175 hours

### 3. GraphQL Support
Related Issue - https://github.com/foss42/apidash/issues/117

GraphQL, a query language for APIs and a runtime for executing those queries, has gained popularity due to its flexibility and efficiency in fetching data. Testing GraphQL APIs is critical for validating schema consistency, query correctness & error handling. By providing support for testing, visualization & integration code generation of GraphQL APIs, we can ensure their quality and reliability.

**Skills**: Understanding Specs, Dart, Flutter
**Difficulty**: Medium-High
**Length**: 90 hours

### 4. gRPC Support
Related Issue - https://github.com/foss42/apidash/issues/14

gRPC (Remote Procedure Call) is an open-source RPC framework developed by Google that facilitates efficient communication between distributed systems. It uses Protocol Buffers (protobuf) as its interface definition language (IDL) and offers features such as bi-directional streaming, authentication, and built-in support for load balancing and health checking. In this project you will add support for testing, visualization & integration code generation gRPC APIs to ensure their functionality, reliability, and performance. 

**Skills**: Understanding Specs, Dart, Flutter
**Difficulty**: Medium-High
**Length**: 90 hours

### 5. Setting up build process for various Linux packaging formats
Related Issue - https://github.com/foss42/apidash/issues/93

AppImage, Flatpak, PKGBUILD, Snap, etc. are different packaging formats or systems designed to simplify the distribution and installation of software on Linux systems, each with its own characteristics and use cases. They provide solutions to common challenges faced by Linux users and developers, such as dependency management, compatibility across distributions, and software isolation. The objective of this project is to setup & document the packaging pipeline so as to make API Dash available to all Linux users.

You can check out our well documented existing build process [here](https://medium.com/@fluttergems/packaging-and-distributing-flutter-desktop-apps-the-missing-guide-part-1-macos-b36438269285). 

**Skills**: Understanding Linux packaging formats, CI/CD, DevOps
**Difficulty**: Low-Medium
**Length**: 90 hours

### 6. Response Previewer for various content types like HTML, CSV, Video, etc. 
Related Issue - https://github.com/foss42/apidash/issues/17 https://github.com/foss42/apidash/issues/155 https://github.com/foss42/apidash/issues/156

API Dash currently supports [40+ mime-types](https://github.com/foss42/apidash?tab=readme-ov-file#mime-types-supported-by-api-dash-response-previewer) and we are looking forward to continue adding support for testing and previewing wide variety of responses such as HTML, CSV, Video, etc. (Check the full IANA list [here](https://www.iana.org/assignments/media-types/media-types.xhtml))

**Skills**: Dart, Flutter
**Difficulty**: Low-Medium
**Length**: 90-175 hours (Based on the no. of mime-types)

### 7. API Testing Suite, Env. Manager, Collection Runner & Monitor
Related Issues - https://github.com/foss42/apidash/discussions/96, https://github.com/foss42/apidash/issues/100, https://github.com/foss42/apidash/issues/25

The objective of this project to design and implement a API testing suite in API Dash so that users can:
- Automate the testing & validation of responses received from an API against predefined expectations (assertions). This will ensure that the API behaves as expected, returns the correct data, and adheres to specified standards and requirements. Users should be able to easily setup automated testing of APIs or API workflows (Chained Requests), execute (automated/manual) the defined test suites and monitor the results.
- Execute a collection of API requests and tests in a systematic and automated manner. This feature is known as Collection Runner, where users can define a sequence of API requests, specify test scripts to be executed against each request, and run the entire collection with a single command. Collection Runner provides batch execution, environment management, data-driven testing, result reporting.

**Skills**: UI/UX Design, Automation, Dart, Flutter
**Difficulty**: Medium-High
**Length**: 350

### 8. Code Generators
Related Discussion - https://github.com/foss42/apidash/discussions/80 [Issue List](https://github.com/foss42/apidash/labels/codegen)

Currently, API Dash supports code generation for the [following language/library](https://github.com/foss42/apidash/tree/main#code-generators). The objective of this project is to add code generators for new languages/libraries/frameworks. Also, special focus will be on generating Frontend UI codes (Flutter, React) so that developers can directly integrate APIs along with the corresponding response visualization in the frontend of their apps.

**Skills**: Dart, Flutter
**Difficulty**: Low-Medium
**Length**: 175

### 9. API Workflow Builder
Related Issue - https://github.com/foss42/apidash/issues/120

The objective of this project is to build an API Workflow Builder in API Dash to simplify API testing and automation by offering a visual interface for creating, managing, and executing test workflows. With drag-and-drop functionality, users can design complex scenarios involving multiple API requests, data transformations, and validations without writing extensive code. The tool should include the capability of dynamic data handling and iteration over datasets for the test workflows to run at scale. Comprehensive monitoring and reporting capabilities should also be included to provide insights into test execution results and performance metrics, aiding in analysis and decision-making.

**Skills**: UI/UX Design, Dart, Flutter
**Difficulty**: Medium-High
**Length**: 350 hours

### 10. Improving UI & Integration Tests
Related Issues - https://github.com/foss42/apidash/issues/118 https://github.com/foss42/apidash/issues/119

The objective of this project is to increase the test coverage of API Dash. This will include adding more unit tests, widget tests, test cases with state management and flows. Also, you will get a good opportunity to design & implement end-to-end integration tests for API Dash.

**Skills**: Testing, Dart, Flutter
**Difficulty**: Low-Medium
**Length**: 175 hours

Please feel free to comment more ideas below👇

**Comment by opxdelwin** [comment]
Hey maintainers,

As there happen to be approximately 10+ https://github.com/foss42/apidash/labels/import-export issues for importing from various file formats like RAML, WADL, cURL, and more, I propose a GSoC project to add comprehensive support for multiple import-export operations, improving interoperability and usability. This project, ranging from 175 to 350 hours, would greatly benefit the project and community.

I am enthusiastic about the opportunity to implement these features throughout the duration of GSoC '24.

Your consideration is greatly appreciated.

Best regards,

**Reply by animator** [reply]
Maybe we can add it later. But, currently other projects have a higher priority.

**Comment by PCoder23** [comment]
Hey team,

I've been thinking, since we don't have workplaces in API Dash yet, why not introduce a collaborative workspace feature? It would involve creating workspaces, enabling collaboration with others, incorporating version control, providing development guides, and optionally adding a chat feature and user permissions. What are your thoughts on this idea?

**Reply by animator** [reply]
Thanks for adding this idea to the list. Adding workspaces is a good feature, but our current focus will be on improving/adding new core features as listed above.

**Comment by ashutosh7i** [comment]
Hello @ashitaprasad & @animator 👋,

My name is Aashutosh, and I am a pre-final year CS student from Indore, India. Professionally, I specialize in Full-Stack Development using JavaScript, and I have experience with Flutter as well.

Personally, am very excited about the project ideas
        2. Realtime API Testing Support - WebSocket, SSE, MQTT and
        3. GraphQL Support
in fact i have made several projects using these technologies, and as an IoT enthusiast, I have extensively used MQTT.

While I primarily use JavaScript, I have also used Dart for a few applications, although they contained minimal backend code compared to my JavaScript projects.
Upon reviewing Apidash's codebase, I noticed it uses Dart.
I am confident that after a week of using Dart, I will become comfortable with it as the concepts are quite similar to what I already know. 
i just need some assurance that i can be a fit for this project,
also please help me out with some tasks or mini projects to get my command in Dart for not just UI.

Thanks and regards,
Aashutosh Soni

**Reply by ashitaprasad** [reply]
Welcome to the project @ashutosh7i 
To get familiar with the codebase, pick up and issue, work on it and send across a PR.

**Reply by ashutosh7i** [reply]
Hello mam,
i have solved issue no https://github.com/foss42/apidash/issues/143 and added codegen support for php guzzle, 
currently i have not created a PR but code can be viewed at this [here](https://github.com/ashutosh7i/apidash/tree/main)
now for the next stage can you suggest me how can i write tests and how to run tests.

**Reply by animator** [reply]
@ashutosh7i Please create a PR for any review.

**Reply by ashutosh7i** [reply]
ok sir,
i have create a PR [here](https://github.com/foss42/apidash/pull/256)

**Comment by yashpandey06** [comment]
Hey There , @ashitaprasad & @animator 👋,

My name is Yash Pandey , and I am a undergraduate student from Indian Institute Of Information Technology .I have experience  in Full-Stack Development using MERN , have also worked with remote startups to make end user applications (from development  to deployment ) based on  SAAS , Python backend frameworks,  UI Frameworks  . I am also a technical content writer who loves to document while developing . 
Though I have just started working in Dart and Flutter , but  I am pretty much sure about my grasping capabilities  and would like to contribute to this awesome  project.

I am very excited about :
10. Improving UI & Integration Tests
8. Code Generators

I have mostly used NodeJS , Python backend .I am surely confident  of using Dart, the concepts are similar and easy to learn, this is the best opportunity to learn while working.
I just need your guidance and assurance to help me out in minimal tasks.
Looking forward to the healthy contribution.

Thanks 

Yash Pandey

**Reply by animator** [reply]
Glad to have you onboard @yashpandey06 🚀

**Comment by lamergameryt** [comment]
Hi @ashitaprasad & @animator,

I'm interested in working on
> 5. Setting up build process for various Linux packaging formats

I need clarification as to what the end goal of the idea is and what deliverables are expected from me at the end of the 90-hour timeline. From my understanding, the goal is to:

- Create a CI/CD pipeline (using GitHub actions?) for automating the creation of a deb, pkgbuild, and rpm distributable file.
- Work on issue [#93](#93): Package the application as a Flatpak for a more streamlined installation across multiple distributions.
- Package the application as a snap for distribution via the Snap Store.

Are there any other additions that need to be made other than the above?

Looking forward to contributing to this amazing project.

Thanks!

**Reply by animator** [reply]
Hi @lamergameryt You can raise an issue with the points mentioned in this discussion and we can start discussing it in the issue thread in detail.

**Reply by lamergameryt** [reply]
Hello @animator, I've created an issue thread #213 for further discussion. Looking forward to hearing your feedback.

**Comment by DILPREET1910** [comment]
Hi @ashitaprasad & @animator!
Dilpreet here.

I'd like to work on:
> 1. Android/iOS support

As per my understanding the main goal of this project is to create a mobile client for android, ios and IPads:
1. while keeping a fairly similar UX to the desktop client.
    - I have gone through the [mock-ups](https://github.com/foss42/apidash/issues/33#issuecomment-1605874012) provided by @animator. I feel they are quite similar to the desktop client and would have a good user experience considering the limited screen space.
    - To begin with we can start coding the mobile client referring this mock-up. And any potential UI/UX enhancements deemed beneficial can be incorporated during the development and testing phases.
    - What are our takes on this?
2. Developing a mobile client in a manner that ensures modifications to functionalities, do not necessitate changes in both the desktop and mobile clients.
3. Rigorous testing of the application ensuring optimal functionality and user experience.

Are there any other requirements I might have overlooked?

Also I went through the codebase and have some quires as to how do  you want me to develop the mobile client in accordance to the `2`nd point. They are section specific.
Should I ask them in the issue [#33](https://github.com/foss42/apidash/issues/33).
Or if you agree on point `1` I can create a new issue and we can start discussing implementation in detail in that thread.

Looking forward to working on this interesting project!!!
Thanks.

**Reply by animator** [reply]
Hi @DILPREET1910 You can ask all your queries in #33

**Comment by UtsavMandal2022** [comment]
Hi @animator and @ashitaprasad 👋,

My name is Utsav Mandal, I am a third year undergraduate student from IIT Kharagpur. I am a Fullstack developer and have a vast experience with the full cycle of development to deployment. I have experience in Backend development with Python, Golang and Node. I also have some relevant experience with Dart and Flutter.

Personally I am more interested towards,
4. gRPC Support
9. API Workflow Builder
6. Response Previewer for various content types like HTML, CSV, Video, etc.

I feel I have more work done on them previously, ensuring smooth start also I can learn a lot while doing the work 😊. 
Looking forward to make great contribution.

Regards,
Utsav Mandal

**Reply by ashitaprasad** [reply]
Hi @UtsavMandal2022 Glad to hear about the projects you are interested in. Feel free to look into issues & send across a PR.

**Comment by monstajoe2002** [comment]
Hello maintainers 👋,
My name is Youssef Amr, a senior Computer Science student at the German International University in Egypt. I'm currently interning at Banque Du Caire as a software engineer, experienced in React, Typescript, Javascript in the web development domain, and working knowledge of Java, Flutter, Python and Rust. 

I'm interested in ideas no. 6 and 8 about response previewer and code generators respectively. As I mentioned above, I am confident in my Flutter skills to be able to contribute in this project for GSoC. What do I need to do to get into GSoC in this project?

Thanks!

**Reply by ashitaprasad** [reply]
Hi @monstajoe2002 Glad to hear about the projects you are interested in. Feel free to look into the related issues & send across a PR.

**Comment by gautamb02** [comment]
Hello @ashitaprasad and @animator !!

This side Gautam Bhagat from Pune.
I've got two ideas which I strongly believe can be useful and impactful.
1. Creating a web application of API Dash for Testing APIs
2. Creating a VS Code extension of API Dash so that is is accessible to developers while coding itself

**Reply by ashitaprasad** [reply]
Hi @gautam-bhagat VS Code extension idea looks cool. Can you create a vscode extension for the example flutter counter app (available on dartpad samples) and share it with us.

**Reply by gautamb02** [reply]
@ashitaprasad Sure!

I've Mid Term currently ongoing which will end on 9 March. Till then I'll gather some resources and start the implementation by 9th of March !

**Reply by animator** [reply]
cool @gautam-bhagat

**Reply by gautamb02** [reply]
Hey @ashitaprasad and @animator !!

I've build a Counter Demo VS Code Extension for use.
The language used was Typescript + Svelte.

Link : [Marketplace Link](https://marketplace.visualstudio.com/items?itemName=GautamBhagat.vsCounter)

Screenshot:
![image](https://github.com/foss42/apidash/assets/73634938/da0f381f-ef6c-42ec-bac1-f8804d6dc347)

After installing the extension, you can see a plus icon on sidebar which will trigger the extension.
I've a very nice idea for the API Dash. Do let me know your availability to discuss further

Regards,
Gautam Bhagat.

**Reply by ashitaprasad** [reply]
Hi @gautam-bhagat We already know that an extension can be built using typescript/javascript. Wanted you to explore if it can be done using Flutter.

**Reply by gautamb02** [reply]
While it's technically possible, it's not a common approach due to the complexity and potential performance issues.

What I was thinking is to built the extension using TypeScript/JavaScript which can be an independent product of API Dash

**Reply by ashitaprasad** [reply]
@gautam-bhagat Before getting into API Dash, as you mentioned it is technically possible to build a vs code extension using flutter, I will recommend you to go ahead and build the flutter counter app as a vs code extension. We can then debug any performance issues that you encounter.

**Reply by gautamb02** [reply]
Hey @animator & @ashitaprasad !

I looked for alot of ways for creating extensions for VS CODe using flutter. Although VSCode only supports JAvascript for extension development!

We can leverage Typescript and Svelte to do the same. I think this can create a very good imapact! If you allow I can start working on proposal!

**Reply by animator** [reply]
Currently, we will undertake only Dart/Flutter projects under GSoC.

**Comment by SyedAbdullah58-dev** [comment]
Hello @ashitaprasad , I am a Software Engineer and not a student anymore. I have over 2 years of work experience and I do have good experience in flutter as well. I am looking to contribute to this project, so am I Eligible to contribute as I am not a student?

**Reply by ashitaprasad** [reply]
Hi @SyedAbdullah58-dev I will recommend you to check out the official GSoC website to check eligibility criteria.

**Reply by SyedAbdullah58-dev** [reply]
![image](https://github.com/foss42/apidash/assets/61689698/9eea58d4-dba0-450d-b079-84d86a3e2fc9)
I think this will do!

**Comment by sixtusagbo** [comment]
Hi @ashitaprasad & @animator,

I am a pre-finalist Computer Science student at Nnamdi Azikiwe University, Nigeria (Unizik). I have experience in Dart and Flutter, C, Python (Flask), Javascript (ReactJS, NextJS), Typescript, PHP (Laravel).

I am interested in:
8. Code Generators.
10. Improving UI & Integration tests.

I'm more excited about 10 because I'll improve my testing skills in the process. I have an idea which I believe will be very useful:

* Support for Variables (with different scopes just like in Postman)

**Reply by animator** [reply]
Welcome to the repo @sixtusagbo 👍 Looking forward to your contributions.

**Comment by mmjsmohit** [comment]
Hi @animator @ashitaprasad!
Would like to suggest the addition of stress testing as a part of API Performance Testing somewhere in this idea list. 
We should be able to:

- Simulate multiple, parallel users that send requests to your APIs continuously for a specified duration.
- See metrics related to the success rate and get benchmark times for a specified endpoint.
- Drilling down into failed requests and their analysis.

**Reply by ashitaprasad** [reply]
This will come under idea number 7

**Comment by morpheus-30** [comment]
Hi @ashitaprasad & @animator ,

I am Nakshatra, a Third year Computer Science student at J.C. Bose University, YMCA, Faridabad. I have experience in Dart and Flutter, C/C++, Python (Flask), Javascript, ReactJS.

As I've previously contributed in codegen last year, I'd be interested in migrating apidash for other platforms
-  Android/IOS support is must.
-  A VSCode integration be an useful integration as we can code and test side by side.
-  Even a chrome extension+website would also come in handy for a lot of people. A properly optimized website, but maybe flutter comes with its performance issues in case of web.

With previously worked with you guys, I am super excited for continuing the same.

**Reply by ashitaprasad** [reply]
Hi @morpheus-30, Great to see you back here 🙌. Eagerly looking forward to your contributions.

**Comment by devc007** [comment]
Hey community!
Devesh chouhan here I was contributing to jenkins in this year for GSoC but as the projects out for which I was preparing. I was hopeless but found this repo. My tech skill is related to this repo project 5  Setting up build process for various Linux packaging formats. I have a decent knowledge of the required skills. I know i am too late but can project maintainer tell me how can i proceed?

**Reply by animator** [reply]
@devc007 To get started, you can go ahead and send across a PR on some open issues.

**Comment by alainjr10** [comment]
Hey @animator .
I'd love to work on idea number 7 for GSOC 24 and beyond.

**Comment by vednig** [comment]
Hi there,
I am final-year student at VIT Bhopal. This will be my first time contributing under GSOC
Looking forward to working on this project by improving Linux Support and  implementing gRPC Support.  Please let me know what anything I can start with? 
I am also Linux user by day and also have experience developing in Dart, Kotlin and Python primarily.

**Reply by vednig** [reply]
I am fairly new to gRPC⚡ and other API standards. Though, for implementing gRPC ⚡Support I would have to read the documentation as. But I believe I can implement it with your support.

**Reply by animator** [reply]
Since the application date is concluding in a couple of days .. I recommend you to directly submit a proposal in the GSoC portal.
